File: ios/RailBridgeIOS/RailBridgeIOS/Bridge/BridgeRequestCoordinator.swift
import Foundation

struct InFlightRequestRecord: Codable, Equatable {
    let correlationId: String
    let method: String
    let callbackId: String?
    let state: String
    let scenario: String
    let startedAt: String
    let elapsedMs: Int
}

final class BridgeRequestCoordinator {
    enum TerminalState: String, Codable {
        case pending = "pending"
        case success = "success"
        case error = "error"
        case timedOut = "timed_out"
        case abandoned = "abandoned"
    }

    struct RegisteredRequest: Equatable {
        let correlationId: String
        let timeoutAtMs: Int64
    }

    struct Decision: Equatable {
        let accepted: Bool
        let terminalState: TerminalState?
        let reason: String
        let terminalAlreadyReached: Bool
        let retryCount: Int

        static func accepted(_ terminalState: TerminalState, retryCount: Int) -> Decision {
            Decision(
                accepted: true,
                terminalState: terminalState,
                reason: "accepted",
                terminalAlreadyReached: false,
                retryCount: retryCount
            )
        }

        static func ignored(
            _ terminalState: TerminalState?,
            reason: String,
            terminalAlreadyReached: Bool,
            retryCount: Int
        ) -> Decision {
            Decision(
                accepted: false,
                terminalState: terminalState,
                reason: reason,
                terminalAlreadyReached: terminalAlreadyReached,
                retryCount: retryCount
            )
        }
    }

    private final class RequestState {
        let correlationId: String
        let method: String
        let callbackId: String?
        let scenario: String
        let startedAt: String
        let startedAtMs: Int64
        let timeoutAtMs: Int64

        var terminalState: TerminalState = .pending
        var retryCount = 0

        init(
            correlationId: String,
            method: String,
            callbackId: String?,
            scenario: String,
            startedAt: String,
            startedAtMs: Int64,
            timeoutAtMs: Int64
        ) {
            self.correlationId = correlationId
            self.method = method
            self.callbackId = callbackId
            self.scenario = scenario
            self.startedAt = startedAt
            self.startedAtMs = startedAtMs
            self.timeoutAtMs = timeoutAtMs
        }
    }

    private let queue = DispatchQueue(label: "com.demo.railbridge.ios.request-coordinator")
    private let timeoutMs: Int64
    private let nowProvider: () -> Int64
    private var requests: [String: RequestState] = [:]

    init(timeoutMs: Int64, nowProvider: @escaping () -> Int64 = { Timestamp.nowMs() }) {
        self.timeoutMs = max(1, timeoutMs)
        self.nowProvider = nowProvider
    }

    func begin(
        correlationId: String,
        method: String,
        callbackId: String?,
        scenario: String,
        startedAt: String
    ) -> RegisteredRequest {
        queue.sync {
            let startedAtMs = nowProvider()
            let state = RequestState(
                correlationId: correlationId,
                method: method,
                callbackId: callbackId,
                scenario: scenario,
                startedAt: startedAt,
                startedAtMs: startedAtMs,
                timeoutAtMs: startedAtMs + timeoutMs
            )
            requests[correlationId] = state
            return RegisteredRequest(correlationId: correlationId, timeoutAtMs: state.timeoutAtMs)
        }
    }

    func acceptSuccess(correlationId: String, retryCount: Int) -> Decision {
        transition(correlationId: correlationId, target: .success, retryCount: retryCount)
    }

    func acceptError(correlationId: String, retryCount: Int) -> Decision {
        transition(correlationId: correlationId, target: .error, retryCount: retryCount)
    }

    func acceptTimeout(correlationId: String) -> Decision {
        queue.sync {
            guard let request = requests[correlationId] else {
                return .ignored(nil, reason: "missing", terminalAlreadyReached: false, retryCount: 0)
            }

            guard request.terminalState == .pending else {
                return .ignored(
                    request.terminalState,
                    reason: "already_terminal",
                    terminalAlreadyReached: true,
                    retryCount: request.retryCount
                )
            }

            guard nowProvider() >= request.timeoutAtMs else {
                return .ignored(.pending, reason: "not_due", terminalAlreadyReached: false, retryCount: request.retryCount)
            }

            request.terminalState = .timedOut
            return .accepted(.timedOut, retryCount: request.retryCount)
        }
    }

    func abandonAll() -> [InFlightRequestRecord] {
        queue.sync {
            var abandoned: [InFlightRequestRecord] = []
            for request in requests.values where request.terminalState == .pending {
                request.terminalState = .abandoned
                abandoned.append(record(from: request))
            }
            return abandoned.sorted(by: { $0.startedAt < $1.startedAt })
        }
    }

    func snapshot() -> [InFlightRequestRecord] {
        queue.sync {
            requests.values
                .filter { $0.terminalState == .pending }
                .map(record(from:))
                .sorted(by: { $0.startedAt < $1.startedAt })
        }
    }

    private func transition(correlationId: String, target: TerminalState, retryCount: Int) -> Decision {
        queue.sync {
            guard let request = requests[correlationId] else {
                return .ignored(nil, reason: "missing", terminalAlreadyReached: false, retryCount: retryCount)
            }

            guard request.terminalState == .pending else {
                return .ignored(
                    request.terminalState,
                    reason: "already_terminal",
                    terminalAlreadyReached: true,
                    retryCount: request.retryCount
                )
            }

            request.terminalState = target
            request.retryCount = max(0, retryCount)
            return .accepted(target, retryCount: request.retryCount)
        }
    }

    private func record(from request: RequestState) -> InFlightRequestRecord {
        InFlightRequestRecord(
            correlationId: request.correlationId,
            method: request.method,
            callbackId: request.callbackId,
            state: request.terminalState.rawValue,
            scenario: request.scenario,
            startedAt: request.startedAt,
            elapsedMs: max(0, Int(nowProvider() - request.startedAtMs))
        )
    }
}

File: app/src/main/java/com/demo/railbridge/bridge/BridgeRequestCoordinator.java
package com.demo.railbridge.bridge;

import com.demo.railbridge.sdk.SdkErrorCode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class BridgeRequestCoordinator {

    private final Map<String, RequestState> requests = new ConcurrentHashMap<>();
    private final Clock clock;
    private final long timeoutMs;

    public BridgeRequestCoordinator(long timeoutMs) {
        this(timeoutMs, Clock.system());
    }

    BridgeRequestCoordinator(long timeoutMs, Clock clock) {
        this.timeoutMs = Math.max(1L, timeoutMs);
        this.clock = clock == null ? Clock.system() : clock;
    }

    public RegisteredRequest begin(
            String correlationId,
            String method,
            String callbackId,
            String scenario,
            String startedAt
    ) {
        RequestState state = new RequestState(
                correlationId,
                method,
                callbackId,
                scenario,
                startedAt,
                clock.nowMs(),
                clock.nowMs() + timeoutMs
        );
        requests.put(correlationId, state);
        return new RegisteredRequest(correlationId, state.timeoutAtMs);
    }

    public Decision acceptSuccess(String correlationId, int retryCount) {
        return transition(correlationId, TerminalState.SUCCESS, retryCount, null);
    }

    public Decision acceptError(String correlationId, int retryCount, SdkErrorCode errorCode) {
        return transition(correlationId, TerminalState.ERROR, retryCount, errorCode);
    }

    public Decision acceptTimeout(String correlationId) {
        RequestState requestState = requests.get(correlationId);
        if (requestState == null) {
            return Decision.ignored(null, "missing", false, 0);
        }
        synchronized (requestState) {
            if (requestState.terminalState != TerminalState.PENDING) {
                return Decision.ignored(requestState.terminalState, "already_terminal", true, requestState.retryCount);
            }
            if (clock.nowMs() < requestState.timeoutAtMs) {
                return Decision.ignored(TerminalState.PENDING, "not_due", false, requestState.retryCount);
            }
            requestState.terminalState = TerminalState.TIMED_OUT;
            requestState.retryCount = Math.max(0, requestState.retryCount);
            return Decision.accepted(TerminalState.TIMED_OUT, requestState.retryCount);
        }
    }

    public List<InFlightRequestRecord> abandonAll() {
        List<InFlightRequestRecord> abandoned = new ArrayList<>();
        for (RequestState requestState : requests.values()) {
            synchronized (requestState) {
                if (requestState.terminalState == TerminalState.PENDING) {
                    requestState.terminalState = TerminalState.ABANDONED;
                    abandoned.add(toRecord(requestState));
                }
            }
        }
        sortByStartedAt(abandoned);
        return abandoned;
    }

    public List<InFlightRequestRecord> snapshot() {
        List<InFlightRequestRecord> inFlight = new ArrayList<>();
        for (RequestState requestState : requests.values()) {
            synchronized (requestState) {
                if (requestState.terminalState == TerminalState.PENDING) {
                    inFlight.add(toRecord(requestState));
                }
            }
        }
        sortByStartedAt(inFlight);
        return inFlight;
    }

    private Decision transition(
            String correlationId,
            TerminalState target,
            int retryCount,
            SdkErrorCode errorCode
    ) {
        RequestState requestState = requests.get(correlationId);
        if (requestState == null) {
            return Decision.ignored(null, "missing", false, retryCount);
        }

        synchronized (requestState) {
            if (requestState.terminalState != TerminalState.PENDING) {
                return Decision.ignored(
                        requestState.terminalState,
                        "already_terminal",
                        true,
                        requestState.retryCount
                );
            }
            requestState.terminalState = target;
            requestState.retryCount = Math.max(0, retryCount);
            requestState.lastErrorCode = errorCode;
            return Decision.accepted(target, requestState.retryCount);
        }
    }

    private InFlightRequestRecord toRecord(RequestState requestState) {
        return new InFlightRequestRecord(
                requestState.correlationId,
                requestState.method,
                requestState.callbackId,
                requestState.terminalState.getValue(),
                requestState.scenario,
                requestState.startedAt,
                Math.max(0L, clock.nowMs() - requestState.startedAtMs)
        );
    }

    private static void sortByStartedAt(List<InFlightRequestRecord> records) {
        Collections.sort(records, new Comparator<InFlightRequestRecord>() {
            @Override
            public int compare(InFlightRequestRecord left, InFlightRequestRecord right) {
                return left.getStartedAt().compareTo(right.getStartedAt());
            }
        });
    }

    interface Clock {
        long nowMs();

        static Clock system() {
            return new Clock() {
                @Override
                public long nowMs() {
                    return System.currentTimeMillis();
                }
            };
        }
    }

    public static final class RegisteredRequest {
        private final String correlationId;
        private final long timeoutAtMs;

        private RegisteredRequest(String correlationId, long timeoutAtMs) {
            this.correlationId = correlationId;
            this.timeoutAtMs = timeoutAtMs;
        }

        public String getCorrelationId() {
            return correlationId;
        }

        public long getTimeoutAtMs() {
            return timeoutAtMs;
        }
    }

    public static final class Decision {
        private final boolean accepted;
        private final TerminalState terminalState;
        private final String reason;
        private final boolean terminalAlreadyReached;
        private final int retryCount;

        private Decision(
                boolean accepted,
                TerminalState terminalState,
                String reason,
                boolean terminalAlreadyReached,
                int retryCount
        ) {
            this.accepted = accepted;
            this.terminalState = terminalState;
            this.reason = reason;
            this.terminalAlreadyReached = terminalAlreadyReached;
            this.retryCount = retryCount;
        }

        static Decision accepted(TerminalState terminalState, int retryCount) {
            return new Decision(true, terminalState, "accepted", false, retryCount);
        }

        static Decision ignored(
                TerminalState terminalState,
                String reason,
                boolean terminalAlreadyReached,
                int retryCount
        ) {
            return new Decision(false, terminalState, reason, terminalAlreadyReached, retryCount);
        }

        public boolean isAccepted() {
            return accepted;
        }

        public TerminalState getTerminalState() {
            return terminalState;
        }

        public String getReason() {
            return reason;
        }

        public boolean isTerminalAlreadyReached() {
            return terminalAlreadyReached;
        }

        public int getRetryCount() {
            return retryCount;
        }
    }

    public enum TerminalState {
        PENDING("pending"),
        SUCCESS("success"),
        ERROR("error"),
        TIMED_OUT("timed_out"),
        ABANDONED("abandoned");

        private final String value;

        TerminalState(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    private static final class RequestState {
        private final String correlationId;
        private final String method;
        private final String callbackId;
        private final String scenario;
        private final String startedAt;
        private final long startedAtMs;
        private final long timeoutAtMs;

        private TerminalState terminalState = TerminalState.PENDING;
        private int retryCount;
        private SdkErrorCode lastErrorCode;

        private RequestState(
                String correlationId,
                String method,
                String callbackId,
                String scenario,
                String startedAt,
                long startedAtMs,
                long timeoutAtMs
        ) {
            this.correlationId = correlationId;
            this.method = method;
            this.callbackId = callbackId;
            this.scenario = scenario;
            this.startedAt = startedAt;
            this.startedAtMs = startedAtMs;
            this.timeoutAtMs = timeoutAtMs;
        }
    }
}
